File: LeetCode/0020_ValidParentheses/Solution.java
public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stk = new Stack<Character>();
        
        for(int i = 0; i<s.length(); i++){
            if(s.charAt(i) == ')'){
                if(stk.size() != 0 && stk.peek() == '('){
                    stk.pop();
                    continue;
                }
                else
                    return false;
            }
            
            if(s.charAt(i) == ']'){
                if(stk.size() != 0 && stk.peek() == '['){
                    stk.pop();
                    continue;
                }
                else
                    return false;
            }

            if(s.charAt(i) == '}'){
                if(stk.size() != 0 && stk.peek() == '{'){
                    stk.pop();
                    continue;
                }
                else
                    return false;
            }
            stk.push(s.charAt(i));
        }
        
        if (stk.size() == 0)
            return true;
        return false;
    }
}

File: LeetCode/0020_ValidParentheses/Solution.py
class Solution:
    # @return a boolean
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack, lookup = [], {"{" : "}", "(" : ")", "[" : "]"}

        for parenthese in s:
            if parenthese in lookup:
                stack.append(parenthese)
            elif len(stack) == 0 or lookup[stack.pop()] != parenthese:
                return False

        return len(stack) == 0

if __name__ == '__main__':
    paren_set1 = "(){}[]{{{}}}"
    paren_set2 = "[{})"
    print(Solution().isValid(paren_set1))
    print(Solution().isValid(paren_set2))
